linera-service/src/linera-exporter/runloops/block_processor.rs

```diff
@@ -0,0 +1,461 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashSet, future::IntoFuture, sync::Arc, time::Duration};
+
+use linera_base::crypto::CryptoHash;
+use linera_chain::types::{Block, ConfirmedBlockCertificate};
+use linera_storage::Storage;
+use tokio::{
+    sync::mpsc::{UnboundedReceiver, UnboundedSender},
+    time::{interval, MissedTickBehavior},
+};
+
+use crate::{
+    common::{BlockId, ExporterError, LiteBlockId},
+    storage::BlockProcessorStorage,
+};
+
+pub(super) struct BlockProcessor<T>
+where
+    T: Storage + Clone + Send + Sync + 'static,
+{
+    storage: BlockProcessorStorage<T>,
+    queue_rear: UnboundedSender<BlockId>,
+    queue_front: UnboundedReceiver<BlockId>,
+}
+
+struct Walker<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    path: Vec<NodeVisitor>,
+    visited: HashSet<BlockId>,
+    storage: &'a mut BlockProcessorStorage<S>,
+}
+
+#[derive(Debug)]
+struct ProcessedBlock {
+    block: BlockId,
+    siblings: Vec<BlockId>,
+    parent: Option<LiteBlockId>,
+}
+
+struct NodeVisitor {
+    node: ProcessedBlock,
+    sibling_branch: usize,
+}
+
+impl<T> BlockProcessor<T>
+where
+    T: Storage + Clone + Send + Sync + 'static,
+{
+    pub(super) fn new(
+        storage: BlockProcessorStorage<T>,
+        queue_rear: UnboundedSender<BlockId>,
+        queue_front: UnboundedReceiver<BlockId>,
+    ) -> Self {
+        Self {
+            storage,
+            queue_rear,
+            queue_front,
+        }
+    }
+
+    pub(super) async fn run_with_shutdown<F>(
+        &mut self,
+        signal: F,
+        persistence_period: u16,
+    ) -> Result<(), ExporterError>
+    where
+        F: IntoFuture<Output = ()>,
+    {
+        let future = signal.into_future();
+        let mut pinned = Box::pin(future);
+
+        let mut interval = interval(Duration::from_secs(persistence_period.into()));
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        loop {
+            tokio::select! {
+
+                biased;
```

Comment:
TIL. Why is the polling order important?

Reply:
Thanks. Actually, I think we should put `self.storage.save()` first. I.e. if these signals: tick to save and shutdown signal are very close to each other, I think we should prioritise persisting block processor's state. If they are far apart (time-wise), we will poll the shutdown signal first anyway.